File: Design/707. Design Linked List/myQueue/myQueue.go
package main

import (
	"errors"
	"fmt"
)

// Node is a structure
type Node struct {
	Value int
	Front *Node
	Back  *Node
}

// MyQueue is a struture contains two sentinel nodes
type MyQueue struct {
	head *Node
	tail *Node
}

// Constructor initializes the queue
func Constructor() MyQueue {
	head := &Node{0, nil, nil}
	tail := &Node{0, nil, nil}
	head.Back = tail
	tail.Front = head

	return MyQueue{head, tail}
}

// Enqueue pushes a node of value val into the back of the queue.
func (q *MyQueue) Enqueue(val int) {
	tail := q.tail
	head := q.tail.Front

	tmp := &Node{val, nil, nil}
	tmp.Front = head
	tmp.Back = tail
	head.Back = tmp
	tail.Front = tmp
}

// Dequeue removes the first node in the front of the queue
func (q *MyQueue) Dequeue() (int, error) {
	if q.head.Back == q.tail {
		return -1, errors.New("empty queue")
	}
	ret := q.head.Back.Value
	q.head.Back = q.head.Back.Back
	return ret, nil
}

func main() {
	q := Constructor()
	data, err := q.Dequeue()
	if err != nil {
		fmt.Println(err)
	}
	q.Enqueue(1)
	q.Enqueue(2)
	q.Enqueue(3)
	data, err = q.Dequeue()
	if err == nil {
		fmt.Println("myQueue Front is: ", data) // return 1
	}
	q.Dequeue()
	fmt.Println("myQueue Front is: ", q.head.Back.Value) // return 3
}

File: Design/707. Design Linked List/C/myQueue.c
#include <stdio.h>
#include <stdlib.h>

typedef struct NODE
{
   int value;
   struct NODE *front;
   struct NODE *back;
} node;

node *head = NULL;
node *tail = NULL;

void createMyQueue() {
    head = calloc(1, sizeof(node));
    tail = calloc(1, sizeof(node));
    head->value = 0;
    head->front = NULL;
    head->back = tail;
    tail->value = 0;
    tail->front = head;
    tail->back = NULL;
}

void enqueue(int value) {
    node *front = tail->front;
    node *tmp = calloc(1, sizeof(node));
    tmp->value = value;
    front->back = tmp;
    tail->front = tmp;
    tmp->front = front;
    tmp->back = tail;
}

void dequeue() {
    if (head->back == tail) {
        return;
    }
    node *back = head->back;
    head->back = head->back->back;
    free(back);
}

void deleteMyQueue() {
    node *front = head;
    while(front->back != NULL){
        node *back = front->back;
        free(front);
        front = back;
    }
}

int main() {
    createMyQueue();
    dequeue();
    enqueue(1);
    enqueue(2);
    enqueue(3);
    dequeue();
    dequeue();
    printf("myQueue Front is: %d \n", head->back->value);
    deleteMyQueue();
}